os, osd: bring the lightweight OMAP iteration

rzarzynski opened the pull request:
<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

rzarzynski commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -7755,28 +7755,32 @@ int PrimaryLogPG::do_osd_ops(OpContext *ctx, vector<OSDOp>& ops)
+	      if (key.substr(0, filter_prefix.size()) != filter_prefix) {
+	        return ObjectStore::omap_iter_ret_t::STOP;
+	      }
+	      if (num >= max_return || bl.length() >= max_bytes) {
```
The limit is known a priori. Perhaps the interface could be extended to accommodate that and allow readahead / minimize size of the critical section under the `Collection::lock`. Yet, RocksDB is only about iterating one-by-one (no batching AFAIK), so rather not for now.

rzarzynski commented on src/kv/KeyValueDB.h:
```diff
@@ -205,6 +206,7 @@ class KeyValueDB {
       return "";
     }
     virtual ceph::buffer::list value() = 0;
+    virtual std::string_view value_as_sv() = 0;
```
Would be nice to kill old variants ultimately. Please note that doing the `memcpy` here results in more fragmentation of `bufferlist` at the `Message` layer.

aclamk replied:
Keys should undergo exactly the same expansion procedure.

There is a problem with `string_view` lifetime. It should be clearly stated for how long the value existence is guaranteed. 
For RocksDB that would be until next `seek/bound/next` call, I am reasonably sure.
For MemStore and KStore it should be checked.

mkogan1 commented:
@rzarzynski have not debugged yet, but with this PR `s3cmd ls ...` always returns an empty LIST
(the LIST op is successful (_200 OK_) but no items are returned)
![image](https://github.com/user-attachments/assets/53c50d71-1c46-4a37-94f4-83fa6bb91822)

rzarzynski commented:
Apologizes, @mkogan1! `rados -p test listomapkeys benchmark_data_o08_3272042_object0` I used in manual testing triggers another flow than RGW / `rados bench`. Fixed the problem and repushed. 

I also added two new commits that bring the latency logging we have in `get_omap_iterator()` / `next()` to make this apple-to-apple. Although `perf` testifies it's pretty costly (~10%), the overall gain looks promising even with those extras.

rzarzynski commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -13727,6 +13734,94 @@ ObjectMap::ObjectMapIterator BlueStore::get_omap_iterator(
+  o->get_omap_tail(&tail);
+  while (it->valid()) {
+    std::string user_key;
+    if (const auto& db_key = it->raw_key().second; db_key >= tail) {
```
`rocksdb::Slice::ToString()` sounds promising. Perhaps it's another place `std::string_view` could help to pacify.

```
                              - 9.46% _ZN14CFIteratorImpl7raw_keyB5cxx11Ev                                                                                                                                        ▒
                                 - 7.32% _ZNK7rocksdb5Slice8ToStringB5cxx11Eb                                                                                                                                     ▒
                                    - _ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE10_M_replaceEmmPKcm                                                                                                    ▒
                                       - _ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE9_M_mutateEmmPKcm                                                                                                   ▒
                                            2.69% _Znwm@plt                                                                                                                                                       ▒
```

rzarzynski replied:
I think this is the leaf of our vast abstraction over the `rocksdb::Iterator`:

```
  string key() override {
    return dbiter->key().ToString();
  } 
  std::pair<std::string, std::string> raw_key() override { 
    return make_pair(prefix, key());
  } 
```

[`rocksdb::Slice` already offers `ToStringView()`](https://github.com/facebook/rocksdb/blob/main/include/rocksdb/slice.h#L95-L97) :-).

rzarzynski commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -7755,34 +7755,41 @@ int PrimaryLogPG::do_osd_ops(OpContext *ctx, vector<OSDOp>& ops)
+	        truncated = true;
+	        return ObjectStore::omap_iter_ret_t::STOP;
+	      }
+	      encode(key, bl);
```
Maybe the `bl` should be initialized with bigger append buffer from the very beginning.

rzarzynski commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -13793,12 +13805,18 @@ int BlueStore::omap_iterate(
     } else {
       ceph_abort();
     }
+  c->store->log_latency(
```
Fixed. Now `perf report` is free of it.

rzarzynski commented on src/kv/RocksDBStore.cc:
```diff
@@ -3027,6 +3046,10 @@ class ShardMergeIteratorImpl : public KeyValueDB::IteratorImpl {
   }
+  std::string_view value_as_sv() override {
+    rocksdb::Slice val = iters[0]->value();
+    return std::string_view{val.data(), val.size()};
```
```suggestion
    return val.ToStringView();
```

rzarzynski commented on src/kv/RocksDBStore.cc:
```diff
@@ -2358,6 +2364,10 @@ class CFIteratorImpl : public KeyValueDB::IteratorImpl {
   }
+  std::string_view value_as_sv() override {
+    rocksdb::Slice val = dbiter->value();
+    return std::string_view{val.data(), val.size()};
```
```suggestion
    return val.ToStringView();
```

rzarzynski commented on src/kv/RocksDBStore.cc:
```diff
@@ -2267,6 +2267,12 @@ bufferptr RocksDBStore::RocksDBWholeSpaceIteratorImpl::value_as_ptr()
 
+std::string_view RocksDBStore::RocksDBWholeSpaceIteratorImpl::value_as_sv()
+{
+  rocksdb::Slice val = dbiter->value();
+  return std::string_view{val.data(), val.size()};
```
```suggestion
  return dbiter->value().ToStringView();
```

rzarzynski commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -7755,34 +7755,41 @@ int PrimaryLogPG::do_osd_ops(OpContext *ctx, vector<OSDOp>& ops)
 	encode(truncated, osd_op.outdata);
 	ctx->delta_stats.num_rd_kb += shift_round_up(osd_op.outdata.length(), 10);
 	ctx->delta_stats.num_rd++;
+	if (result < 0) {
```
Move this up.

rzarzynski commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -13727,6 +13734,93 @@ ObjectMap::ObjectMapIterator BlueStore::get_omap_iterator(
+    if (const auto& db_key = it->raw_key().second; db_key >= tail) {
+      break;
+    } else {
+      o->decode_omap_key(db_key, &user_key);
```
This could easily work with `std::string_view` while being worth around 4% of cycles;

```cpp
void BlueStore::Onode::decode_omap_key(const string& key, string *user_key)
{
  size_t pos = sizeof(uint64_t) + 1;
  if (!onode.is_pgmeta_omap()) {
    if (onode.is_perpg_omap()) {
      pos += sizeof(uint64_t) + sizeof(uint32_t);
    } else if (onode.is_perpool_omap()) {
      pos += sizeof(uint64_t);
    }
  }
  *user_key = key.substr(pos);
}
```

```
                              - 4.53% _ZN9BlueStore5Onode15decode_omap_keyERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEPS6_                                                                             ▒
                                 - _ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE12_M_constructIPKcEEvT_S8_St20forward_iterator_tag                                                                        ▒
                                      1.40% _Znwm@plt                                                                                                                                                             ▒
                                      0.52% memcpy@plt                                                                                                                                                            ▒
```

mkogan1 reviewed (approved):
following the same methodology used to test PR https://github.com/ceph/ceph/pull/60000
S3 LIST of 50M objects

the measured performance improvement is unequivocal:
7 minutes faster with this PR, 
percentage difference - the 50 million object listing duration was 17.26% faster.

```
time (nice numactl -N 1 -m 1 -- ~/go/bin/hsbench -a b2345678901234567890 -s b234567890123456789012345678901234567890 -u http://127.0.0.1:8000 -z 4K -d -1 -t $(( $(numactl -N 0 -- nproc) / 1 )) -b 1 -n 50000000 -m l -bp b01b- -op 'folder01/stage01_')

# Before PR: 47:20.61 total
2024/10/09 08:06:42 Running Loop 0 BUCKET LIST TEST
2024/10/09 08:54:02 Loop: 0, Int: TOTAL, Dur(s): 2840.6, Mode: LIST, Ops: 50000, MB/s: 0.00, IO/s: 18, Lat(ms): [ min: 43.5, avg: 56.8, 99%: 76.6, max: 120.5 ], Slowdowns: 0               
( nice numactl -N 1 -m 1 -- ~/go/bin/hsbench -a b2345678901234567890 -s  -u  )  2163.04s user 85.44s system 79% cpu 47:20.61 total

# After PR: 40:22.58 total
2024/10/14 11:42:08 Running Loop 0 BUCKET LIST TEST 
2024/10/14 12:22:31 Loop: 0, Int: TOTAL, Dur(s): 2422.6, Mode: LIST, Ops: 50000, MB/s: 0.00, IO/s: 21, Lat(ms): [ min: 36.8, avg: 48.5, 99%: 68.5, max: 117.8 ], Slowdowns: 0 
( nice numactl -N 1 -m 1 -- ~/go/bin/hsbench -a b2345678901234567890 -s  -u  )  2140.62s user 76.28s system 91% cpu 40:22.58 total
```

rzarzynski commented:
@mkogan1: do you have an information (`perf stat` preferably) on how many cycles the OSDs burnt to serve those S3 requests?

In [the yesterday's testing](https://gist.github.com/rzarzynski/454e5d31ce0e84333b79bfe88362efdb) I'm seeing around 400% more IOPS:

```
>>> 2702 / 649
4.163328197226503
```

but:

* the environment is controlled for saturation (to ensure any improvement at a `tp_osd_tp` thread directly translates into more IOPS; not just lesser resource utilization);
* the testing happens at the RADOS layer, without RGW. I think it's the biggest difference and what is visible at the S3 layer is reduction in latency of OSD + RGW **combined** (in other words: RGW increases the denominator).

aclamk commented on src/os/ObjectStore.h:
```diff
@@ -766,6 +767,36 @@ class ObjectStore {
+  /**
+   * Iterate object map with user-provided callable
+   *
+   * Warning!  The callable is executed under lock on filestore
```
Referring to filestore would confuse people.

aclamk commented on src/os/ObjectStore.h:
```diff
@@ -766,6 +767,36 @@ class ObjectStore {
+    STOP,
+    NEXT
+  };
+  virtual int omap_iterate(
```
I like the concept of omap iteration as a black box.
It would be so much better if we do not expose RocksDB iterator behaviours to OSD.
Yes, i really propose to delete `OmapIterator` class altogether.

rzarzynski replied:
I'm on board with this! `omap_iter_ret_t` already provides room for e.g. `PREV`; similarly `SEEK_TO_FIRST` could be added to `omap_iter_seek_t`.

Another interface-layer cleanup would be dropping (some of) the `std::string`-centric methods of `KeyValueDB::IteratorImpl` (then hopefully used only by providers of `ObjectStore`).

All of this as follow-ups with the reason being backportability – personally I'm fine even with short-living duplication of the code if it helps minimizing intrusiveness.

rzarzynski replied:
https://github.com/ceph/ceph/pull/60890

aclamk commented on src/os/ObjectStore.h:
```diff
@@ -766,6 +767,36 @@ class ObjectStore {
+    const ghobject_t &oid, ///< [in] object
+    omap_iter_seek_t start_from, ///< [in] where the iterator should point to at the beginning
+    std::function<omap_iter_ret_t(std::string_view, std::string_view)> f
+  ) {
```
There is something missing in this interface that users of OmapIterator could do.
When user of OmapIterator wanted to retireve exactly 30 elements it could check if there is one more.
If not, it returned in message that "it is all"
```
PrimaryLogPG::do_osd_ops()
...
 case CEPH_OSD_OP_OMAPGETVALS:
 ...
 bool truncated;
 ```

rzarzynski replied:
I believe `truncated` is already covered: https://github.com/ceph/ceph/pull/60278/files#diff-04d5a3324b706d64a3230221d2eb08f6aa66829f93ca151ebb71896fb743ada7R7776.

rzarzynski commented on src/os/ObjectStore.h:
```diff
@@ -766,6 +767,36 @@ class ObjectStore {
+    omap_iter_seek_t start_from, ///< [in] where the iterator should point to at the beginning
+    std::function<omap_iter_ret_t(std::string_view, std::string_view)> f
+  ) {
+    return -EOPNOTSUPP;
```
Of course, this is definitely a TODO before.

rzarzynski commented:
@mkogan1: OK, so the listing of constant number (50M) of objects took `2418 / 2744 ~= 0.88` of time while burning `13565614722638 / 19146385188311 ~= 0.70` cycles. Not so bad, I would say :-). Yet, maybe we can squeeze more. Any chance for flamegraphs?

rzarzynski commented:
Oh, report from benchmarking & profiling https://github.com/ceph/ceph/pull/60278/commits/ec6ed2620086eaa3758f0d7a9f6968ad782a8844 is available here: https://gist.github.com/rzarzynski/454e5d31ce0e84333b79bfe88362efdb?permalink_comment_id=5234618#gistcomment-5234618.

mkogan1 commented:
> @mkogan1: OK, so the listing of constant number (50M) of objects took  ... Any chance for flamegraphs?

attached, (without the 2 commits from yesterday yet)
perf was collected as following: `sudo perf record --call-graph 'lbr' -m 8M --aio -z --pid=$(pidof ceph-osd) -- sleep 60`
there are many unknown stacks, will see if can improve
![ceph-osd_583018_lbr_60](https://github.com/user-attachments/assets/e3412d92-64bf-4328-95f8-34dea5150e9a)

mkogan1 commented:
updated flame graph with PR (+ the 2 commits from yesterday), and fixed unknown stacks (the methodology was changed slightly):
```
sudo perf record -e 'cycles' --switch-events --sample-cpu -g --call-graph 'lbr' -m 8M --aio -z --pid=$(pidof ceph-osd) -- sleep 30
time (sudo perf script | stackcollapse-perf.pl --kernel | flamegraph.pl --width 1600 --bgcolors grey --cp --hash > ceph-osd_$(pidof ceph-osd).svg)
```
![ceph-osd_1653738](https://github.com/user-attachments/assets/579babb7-0c72-4e5d-b8d9-bf4a7c746cea)

cbodley commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -13727,6 +13740,101 @@ ObjectMap::ObjectMapIterator BlueStore::get_omap_iterator(
+    }
+    omap_iter_ret_t ret = f(user_key, it->value_as_sv());
+    if (ret == omap_iter_ret_t::STOP) {
+      ceph::time_guard<ceph::mono_clock>{next_lat_acc};
```
i don't think this time_guard is doing what you intended, since it goes out of scope immediately

rzarzynski replied:
You're absolutely right; it should be one clause below. Thank you, Casey!

aclamk replied:
Still not moved.

rzarzynski replied:
Fixed.

rzarzynski commented on src/cls/rgw/cls_rgw.cc:
```diff
@@ -640,19 +631,41 @@ int rgw_bucket_list(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
-        continue;
-      }
+      // these checks are performed also by RGW
+      if (rgw_do_osd_side_checks) {
```
The cost of these checks could be lowered if the flags field in the encoding schema move at the beginning. Yet, this would require using a feature (/ release) bit.

rzarzynski commented on src/cls/rgw/cls_rgw.cc:
```diff
@@ -640,19 +631,41 @@ int rgw_bucket_list(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
+        start_after_entry_key = cls_rgw_obj_key{};
       }
 
       if (has_delimiter) {
```
Hmm, perhaps the straight case (no delimiter, no extensive versioning) of bucket listing could bypass the `cls_rgw` entirely – RGW (or maybe even `cls_rgw_client`) could emit `CEPH_OSD_OP_OMAPGETVALS`  (instead of `CEPH_OSD_OP_CALL`) relying the `is_valid()` and `is_visible()` checks are performed client-side.

The question is how large the _ugly namespace_ of keys can be.

Dynamic switching between the direct and cls-assisted paths is not unthinkable.

mkogan1 commented:
@rzarzynski updated flamegraph - at`* main      24af2ff8925` + 21 commits `git cherry-pick -ex dc31d0dfca70fd2e4be4748caf197d4f59e0211d^..9f0f763aceb1d09392c4ac5c5ed7f85e38ef17641`
of `hsbench` list of 50M obj bucket: `time (nice numactl -N 1 -m 1 -- ~/go/bin/hsbench -a b2345678901234567890 -s b234567890123456789012345678901234567890 -u http://127.0.0.1:8000 -z 4K -d -1 -t $(( $(numactl -N 0 -- nproc) / 1 )) -b 1 -n 50000000 -m l -bp b01b- -op 'folder01/stage01_' |& tee hsbench.log)`

and perf record & perf ... flamegraph.pl:
```
❯ sudo perf record -e 'cycles' --switch-events --sample-cpu -g --call-graph 'lbr' -m 8M --aio -z --pid=$(pidof ceph-osd) -- sleep 30
time (sudo perf script | stackcollapse-perf.pl --kernel | flamegraph.pl --width 1600 --bgcolors grey --cp --hash > ceph-osd_$(pidof ceph-osd).svg)
```

![ceph-osd_1306406](https://github.com/user-attachments/assets/8c2f4b31-72d7-412f-86fa-7b0cdf8089a8)

markhpc commented:
> I bet the (single?) RGW instance isn't able to saturate the OSD. Actually, the flamegrah shows a sign of undersaturation – significant share of cycles spent on thread wait / sleep coordination.

This is possible.  In some testing I was seeing that on fast NVMe drives I needed 1 RGW instance per 3 or even 2 OSDs to maximize performance, and it wouldn't be out of the question that in some extreme cases a single RGW instance may not be able to saturate a single OSD.

athanatos commented:
@yehudasa This PR reworks the internal ObjectStore omap scanning interface.  I think I recall that you were looking for a way to do reverse iteration?  Might be worth weighing in here.

athanatos commented:
This looks very reasonable to me!

rzarzynski commented:
I will dissect the non-OSD parts, particularly the RGW changes, to dedicated PRs and add here a commit deprecating the old `ObjectStore::get_omap_iterator()` interface.

rzarzynski commented:
The RGW and `objclass` commits have been dissected to dedicated PR for reviews.
Fixups got squashed.
The `get_omap_iterator` interface has been deprecated.

Matan-B commented on src/crimson/os/alienstore/alien_store.cc:
```diff
@@ -429,8 +429,21 @@ auto AlienStore::omap_get_values(CollectionRef ch,
+        c->collection, oid,
+        ObjectStore::omap_iter_seek_t{
+          .seek_position = start.value_or(std::string{}),
+          // FIXME: classical OSDs begins iteration from LOWER_BOUND
+          // (or UPPER_BOUND if filter_prefix > start). However, these
+          // bits are not implemented yet
```
Which bits are missing for Alien store? Is it `PrimaryLogPG::do_osd_ops` usage of `omap_iterate`?

rzarzynski replied:
Yes, it has some (tiny) features we miss in crimson like `filter_prefix` support.

Matan-B reviewed (commented):
Crimson changes lgtm. Can you please ping once this PR is ready for testing? so we can make sure nothing breaks on AlienStore side.

rzarzynski commented:
I added the support `MemStore` and `KStore`. Deprecation of `get_omap_iterator()` happened with the WIP being its complete removal (see https://github.com/rzarzynski/ceph/pull/new/wip-os-fastomapiter-killomapiterator).

rzarzynski commented:
@Matan-B, @athanatos: fine to approve?

aclamk commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -4830,7 +4830,7 @@ void BlueStore::Onode::rewrite_omap_key(const string& old, string *out)
 }
 
-void BlueStore::Onode::decode_omap_key(const string& key, string *user_key)
+size_t BlueStore::Onode::get_userkey_offset_in_omap_key() const
```
Can I please get "calc_" instead of "get_" ?
In BS we use this distinction when we retrieve a value that is not a property, but somehow a result of data structures / offsets.

aclamk commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -13727,6 +13694,95 @@ ObjectMap::ObjectMapIterator BlueStore::get_omap_iterator(
+
+  // iterate!
+  std::string tail;
+  o->get_omap_tail(&tail);
```
We already calculated `get_omap_tail` as `upper_bound`.
Maybe we can just exposed it and here just use it?

aclamk commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -13727,6 +13694,95 @@ ObjectMap::ObjectMapIterator BlueStore::get_omap_iterator(
+    o->get_userkey_offset_in_omap_key();
+  ceph::timespan next_lat_acc{0};
+  while (it->valid()) {
+    std::string_view user_key;
```
nit: user_key -> omap_key ?

rzarzynski replied:
I think breaking the current convention would bring more confusion than benefit. Although `omap_key` would IMHO be correct, `omap_get_keys()` & co use `user_key`.

aclamk commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -13727,6 +13694,95 @@ ObjectMap::ObjectMapIterator BlueStore::get_omap_iterator(
+      break;
+    } else {
+      user_key = db_key.substr(userkey_offset_in_dbkey);
+    }
```
NIT: I think having loop-break conditions separate from other codes looks better:
```
const auto& db_key = it->raw_key_as_sv().second;
if (db_key >= tail) {
  break;
}
std::string_view user_key = db_key.substr(userkey_offset_in_dbkey);
```

aclamk reviewed (approved):
Good stuff.
Minor comments only about doc and style.

rzarzynski commented on src/kv/RocksDBStore.cc:
```diff
@@ -2006,6 +2008,27 @@ int RocksDBStore::split_key(rocksdb::Slice in, string *prefix, string *key)
+  if (separator == NULL)
+     return -EINVAL;
+  prefix_len = size_t(separator - in.data());
+  if (prefix_len >= in.size())
```
At first glance only the `==` part of the condition could be meaningful. However, the check for mere occurrence already exists above as:

>        The  memchr()  and memrchr() functions return a pointer to the matching
>        byte or NULL if the character does not occur in the given memory area.

rzarzynski replied:
Completely theoretical justification could be we don't use `std::uintptr_t` for the pointer arithmetic Yet, as those pointer are within same-object boundaries, `size_t` should be enough:

> as std::size_t can store the maximum size of a theoretically possible object of any type (including array).

rzarzynski replied:
Will push a separate commit.

Matan-B reviewed (approved):
crimson/* approved. Can you please share the build here so I'll run it through Crimson's suite?

rzarzynski commented:
Commit https://github.com/ceph/ceph/pull/60278/commits/19392d4c8215e5b933649b40e3670d5c9678a159, being a response to a review comment, will pushed the follow-up PR.

rzarzynski commented:
I'm dropping the `rgw` label as the RGW commits have been dissected to the follow-up PR.

ljflores commented:
Rados approved: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrackercephcomissues69287
